Deleted the bucket object and pod/crossplane-provider-influxdb went into CrashLoopBackOff

<!--
Thank you for helping to improve InfluxDB Cloud Provider!

Please be sure to search for open issues before raising a new one. We use issues
for bug reports and feature requests. Please find us at https://slack.crossplane.io
for questions, support, and discussion.
-->

### What happened?
<!--
Please let us know what behaviour you expected and how InfluxDB Cloud Provider
diverged from that behaviour.
-->

I created a bucket by applying the following manifest:

```
apiVersion: influxdb.crossplane.io/v1alpha1
kind: Bucket
metadata:
  name: example-bucket2
spec:
  forProvider:
    description: example-bucket2
    orgID: SOME_ORG_ID
    retentionRules:
    - everySeconds: 0
      type: expire
  providerConfigRef:
    name: default
```

The bucket got created and was visible on the InfluxDB Cloud 2.0 UI.

Although the resource never got to the state `Ready: True`. See below:
```
$ kubectl get buckets.influxdb.crossplane.io                
NAME              READY   SYNCED   EXTERNAL-NAME     AGE
example-bucket2           True     example-bucket2   5m15s
```

Then I tried to delete with:

```
$ kubectl delete buckets.influxdb.crossplane.io example-bucket2
bucket.influxdb.crossplane.io "example-bucket2" deleted
```

Then the `crossplane-provider-influxdb` pod got into `CrashLoopBackOff`

```
NAME                                                              READY   STATUS             RESTARTS   AGE
pod/crossplane-78bf69f4cc-f2zmz                                   1/1     Running            0          27h
pod/crossplane-provider-azure-ddc1abaec4d0-665f9bb6c8-mplm8       1/1     Running            1          27h
pod/crossplane-provider-influxdb-ba2ab2715129-8567c78666-pz57p    0/1     CrashLoopBackOff   50         3h59m
pod/crossplane-provider-jet-azure-de7fdc8d3f04-84d4874476-dstzb   1/1     Running            0          27h
pod/crossplane-provider-sql-f6f5be714a28-545b559dcb-m8kw4         1/1     Running            0          27h
pod/crossplane-provider-tf-azure-529a2cefbfc7-5ffdf4f86b-8pdcl    1/1     Running            0          27h
pod/crossplane-rbac-manager-86996b55b4-dn57k                      1/1     Running            1          27h
pod/upbound-bootstrapper-65cbb55d7c-f8z54                         1/1     Running            0          27h
pod/xgql-9bbf5d9d7-4f75j                                          1/1     Running            0          27h
```

The following log output was seen on the pod mentioned above:

```
I0112 14:01:23.925955       1 request.go:668] Waited for 1.011662735s due to client-side throttling, not priority and fairness, request: GET:https://192.168.0.1:443/apis/postgresql.azure.tf.crossplane.io/v1alpha1?timeout=32s
panic: runtime error: index out of range [0] with length 0

goroutine 444 [running]:
github.com/crossplane-contrib/provider-influxdb/internal/controller/bucket.LateInitialize(0xc0001b2140, 0xc0000101e0, 0x1)
        /home/runner/work/provider-influxdb/provider-influxdb/internal/controller/bucket/conversions.go:129 +0x35f
github.com/crossplane-contrib/provider-influxdb/internal/controller/bucket.(*external).Observe(0xc0008440b0, 0x1c4cdc0, 0xc000319920, 0x1c6f690, 0xc0001b2000, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/runner/work/provider-influxdb/provider-influxdb/internal/controller/bucket/controller.go:98 +0x476
github.com/crossplane/crossplane-runtime/pkg/reconciler/managed.(*Reconciler).Reconcile(0xc0006c44d0, 0x1c4cdf8, 0xc00092e360, 0x0, 0x0, 0xc0009826e0, 0xf, 0xc00092e300, 0x0, 0x0, ...)
        /home/runner/work/provider-influxdb/provider-influxdb/vendor/github.com/crossplane/crossplane-runtime/pkg/reconciler/managed/reconciler.go:679 +0x22ab
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0003234a0, 0x1c4cd50, 0xc000429540, 0x1879340, 0xc00091b340)
        /home/runner/work/provider-influxdb/provider-influxdb/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:298 +0x30d
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0003234a0, 0x1c4cd50, 0xc000429540, 0x663a226465636100)
        /home/runner/work/provider-influxdb/provider-influxdb/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:253 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2(0xc0003b2010, 0xc0003234a0, 0x1c4cd50, 0xc000429540)
        /home/runner/work/provider-influxdb/provider-influxdb/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:214 +0x6b
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
        /home/runner/work/provider-influxdb/provider-influxdb/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:210 +0x425
```

### How can we reproduce it?
<!--
Help us to reproduce your bug as succinctly and precisely as possible. Artifacts
such as example manifests or a script that triggers the issue are highly
appreciated!
-->
The steps are described in the `what happened?` section.
 
### What environment did it happen in?

<!--
Include at least the version or commit you were running. Consider
also including your:

* Cloud provider or hardware configuration
* Kubernetes version (use `kubectl version`)
* Kubernetes distribution (e.g. Tectonic, GKE, OpenShift)
* OS (e.g. from /etc/os-release)
* Kernel (e.g. `uname -a`)
-->
Crossplane version: upbound/crossplane:v1.5.1-up.1
Provider InfluxDB Cloud version: v0.1.1
Azure AKS
Kubernetes 1.20.13

--- a/apis/v1alpha1/bucket_types.go
+++ b/apis/v1alpha1/bucket_types.go
@@ -46,7 +46,8 @@ type BucketParameters struct {
 	RP *string `json:"rp,omitempty"`
 
 	// Rules to expire or retain data. No rules means data never expires.
-	RetentionRules []RetentionRule `json:"retentionRules"`
+	// +optional
+	RetentionRules []RetentionRule `json:"retentionRules,omitempty"`
 
 	// +kubebuilder:validation:Enum=implicit;explicit
 	SchemaType string `json:"schemaType,omitempty"`

--- a/internal/controller/bucket/controller_test.go
+++ b/internal/controller/bucket/controller_test.go
@@ -112,6 +112,56 @@ func TestObserve(t *testing.T) {
 				},
 			},
 		},
+		"ExpireNoUpdateNeeded": {
+			args: args{
+				mg: &v1alpha1.Bucket{
+					Spec: v1alpha1.BucketSpec{
+						ForProvider: v1alpha1.BucketParameters{
+							Description:    pointer.String("bucket"),
+							RetentionRules: []v1alpha1.RetentionRule{{Type: "expire", EverySeconds: 0}},
+						},
+					},
+				},
+				api: &clients.MockBucketsAPI{
+					FindBucketByNameFn: func(_ context.Context, _ string) (*domain.Bucket, error) {
+						return &domain.Bucket{
+							Description: pointer.String("bucket"),
+						}, nil
+					},
+				},
+			},
+			want: want{
+				obs: managed.ExternalObservation{
+					ResourceExists:   true,
+					ResourceUpToDate: true,
+				},
+			},
+		},
+		"ExpireUpdateNeeded": {
+			args: args{
+				mg: &v1alpha1.Bucket{
+					Spec: v1alpha1.BucketSpec{
+						ForProvider: v1alpha1.BucketParameters{
+							Description:    pointer.String("bucket"),
+							RetentionRules: []v1alpha1.RetentionRule{{Type: "expire", EverySeconds: 3600}},
+						},
+					},
+				},
+				api: &clients.MockBucketsAPI{
+					FindBucketByNameFn: func(_ context.Context, _ string) (*domain.Bucket, error) {
+						return &domain.Bucket{
+							Description: pointer.String("bucket"),
+						}, nil
+					},
+				},
+			},
+			want: want{
+				obs: managed.ExternalObservation{
+					ResourceExists:   true,
+					ResourceUpToDate: false,
+				},
+			},
+		},
 	}
 
 	for name, tc := range cases {

--- a/internal/controller/bucket/conversions.go
+++ b/internal/controller/bucket/conversions.go
@@ -117,33 +117,50 @@ func LateInitialize(params *v1alpha1.BucketParameters, obs *domain.Bucket) bool 
 	params.RP = li.LateInitializeStringPtr(params.RP, obs.Rp)
 	if params.SchemaType == "" && obs.SchemaType != nil && string(*obs.SchemaType) != "" {
 		params.SchemaType = string(*obs.SchemaType)
-		return true
+		li.SetChanged()
 	}
-	sort.SliceStable(params.RetentionRules, func(i, j int) bool {
-		return params.RetentionRules[i].Type < params.RetentionRules[j].Type
-	})
-	sort.SliceStable(obs.RetentionRules, func(i, j int) bool {
-		return string(obs.RetentionRules[i].Type) < string(obs.RetentionRules[j].Type)
-	})
-	for i := range params.RetentionRules {
-		if params.RetentionRules[i].Type == string(obs.RetentionRules[i].Type) {
-			params.RetentionRules[i].ShardGroupDurationSeconds = li.LateInitializeInt64Ptr(params.RetentionRules[i].ShardGroupDurationSeconds, obs.RetentionRules[i].ShardGroupDurationSeconds)
+
+	if len(params.RetentionRules) == 0 && len(obs.RetentionRules) > 0 {
+		sort.SliceStable(obs.RetentionRules, func(i, j int) bool {
+			return string(obs.RetentionRules[i].Type) < string(obs.RetentionRules[j].Type)
+		})
+
+		params.RetentionRules = make([]v1alpha1.RetentionRule, len(obs.RetentionRules))
+		for i := range obs.RetentionRules {
+			params.RetentionRules[i] = v1alpha1.RetentionRule{
+				Type:                      string(obs.RetentionRules[i].Type),
+				EverySeconds:              obs.RetentionRules[i].EverySeconds,
+				ShardGroupDurationSeconds: obs.RetentionRules[i].ShardGroupDurationSeconds,
+			}
 		}
+		li.SetChanged()
 	}
 	return li.IsChanged()
 }
 
 // IsUpToDate returns whether an update call is necessary.
 func IsUpToDate(params v1alpha1.BucketParameters, obs *domain.Bucket) bool {
+	// RetentionRules observation is empty if EverySeconds is set to 0
+	if len(obs.RetentionRules) == 0 {
+		obs.RetentionRules = []domain.RetentionRule{{Type: "expire"}}
+	}
+	// If params have no RetentionRules, assume no expiry
+	if len(params.RetentionRules) == 0 {
+		params.RetentionRules = []v1alpha1.RetentionRule{{Type: "expire"}}
+	}
+
 	if len(params.RetentionRules) != len(obs.RetentionRules) {
 		return false
 	}
+
 	sort.SliceStable(params.RetentionRules, func(i, j int) bool {
 		return params.RetentionRules[i].Type < params.RetentionRules[j].Type
 	})
+
 	sort.SliceStable(obs.RetentionRules, func(i, j int) bool {
 		return string(obs.RetentionRules[i].Type) < string(obs.RetentionRules[j].Type)
 	})
+
 	for i := range params.RetentionRules {
 		if params.RetentionRules[i].Type != string(obs.RetentionRules[i].Type) ||
 			params.RetentionRules[i].EverySeconds != obs.RetentionRules[i].EverySeconds ||
@@ -151,6 +168,7 @@ func IsUpToDate(params v1alpha1.BucketParameters, obs *domain.Bucket) bool {
 			return false
 		}
 	}
+
 	return pointer.StringDeref(obs.Description, "") == pointer.StringDeref(params.Description, "")
 }
 

--- a/internal/controller/bucket/conversions_test.go
+++ b/internal/controller/bucket/conversions_test.go
@@ -1,0 +1,259 @@
+/*
+Copyright 2021 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package bucket
+
+import (
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/influxdata/influxdb-client-go/v2/domain"
+	"k8s.io/utils/pointer"
+
+	"github.com/crossplane-contrib/provider-influxdb/apis/v1alpha1"
+)
+
+func TestLateInitialize(t *testing.T) {
+	sType := domain.SchemaTypeExplicit
+	type args struct {
+		params *v1alpha1.BucketParameters
+		obs    *domain.Bucket
+	}
+	type want struct {
+		params *v1alpha1.BucketParameters
+		res    bool
+	}
+	cases := map[string]struct {
+		args args
+		want want
+	}{
+		"UpToDate": {
+			args: args{
+				params: &v1alpha1.BucketParameters{
+					Description: pointer.String("name"),
+				},
+				obs: &domain.Bucket{
+					Description: pointer.String("name"),
+				},
+			},
+			want: want{
+				params: &v1alpha1.BucketParameters{
+					Description: pointer.String("name"),
+				},
+				res: false,
+			},
+		},
+		"LateInitDescription": {
+			args: args{
+				params: &v1alpha1.BucketParameters{},
+				obs: &domain.Bucket{
+					Description: pointer.String("LIname"),
+				},
+			},
+			want: want{
+				params: &v1alpha1.BucketParameters{
+					Description: pointer.String("LIname"),
+				},
+				res: true,
+			},
+		},
+		"LateInitSchemaType": {
+			args: args{
+				params: &v1alpha1.BucketParameters{},
+				obs: &domain.Bucket{
+					Description: pointer.String("name"),
+					SchemaType:  &sType,
+				},
+			},
+			want: want{
+				params: &v1alpha1.BucketParameters{
+					Description: pointer.String("name"),
+					SchemaType:  "explicit",
+				},
+				res: true,
+			},
+		},
+		"LateInitRetentionRules": {
+			args: args{
+				params: &v1alpha1.BucketParameters{
+					Description: pointer.String("name"),
+				},
+				obs: &domain.Bucket{
+					Description:    pointer.String("LIname"),
+					RetentionRules: []domain.RetentionRule{{Type: "expire"}},
+				},
+			},
+			want: want{
+				params: &v1alpha1.BucketParameters{
+					Description:    pointer.String("name"),
+					RetentionRules: []v1alpha1.RetentionRule{{Type: "expire"}},
+				},
+				res: true,
+			},
+		},
+		"LateInitIgnoreRetentionRules": {
+			args: args{
+				params: &v1alpha1.BucketParameters{
+					Description:    pointer.String("name"),
+					RetentionRules: []v1alpha1.RetentionRule{{Type: "expire", EverySeconds: 3600}},
+				},
+				obs: &domain.Bucket{
+					Description:    pointer.String("LIname"),
+					RetentionRules: []domain.RetentionRule{{Type: "expire"}},
+				},
+			},
+			want: want{
+				params: &v1alpha1.BucketParameters{
+					Description:    pointer.String("name"),
+					RetentionRules: []v1alpha1.RetentionRule{{Type: "expire", EverySeconds: 3600}},
+				},
+				res: false,
+			},
+		},
+	}
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			res := LateInitialize(tc.args.params, tc.args.obs)
+
+			if diff := cmp.Diff(tc.want.res, res); diff != "" {
+				t.Errorf("Observe(...): -want, +got:\n%s", diff)
+			}
+			if diff := cmp.Diff(tc.want.params, tc.args.params); diff != "" {
+				t.Errorf("Observe(...): -want, +got:\n%s", diff)
+			}
+		})
+	}
+}
+func TestIsUpToDate(t *testing.T) {
+	type args struct {
+		params v1alpha1.BucketParameters
+		obs    *domain.Bucket
+	}
+	cases := map[string]struct {
+		args args
+		want bool
+	}{
+		"UpToDate": {
+			args: args{
+				params: v1alpha1.BucketParameters{
+					Description: pointer.String("name"),
+				},
+				obs: &domain.Bucket{
+					Description: pointer.String("name"),
+				},
+			},
+			want: true,
+		},
+		"DescriptionDifferent": {
+			args: args{
+				params: v1alpha1.BucketParameters{
+					Description: pointer.String("name"),
+				},
+				obs: &domain.Bucket{
+					Description: pointer.String("differentName"),
+				},
+			},
+			want: false,
+		},
+		"DefaultExpire": {
+			args: args{
+				params: v1alpha1.BucketParameters{
+					Description: pointer.String("name"),
+				},
+				obs: &domain.Bucket{
+					Description:    pointer.String("name"),
+					RetentionRules: []domain.RetentionRule{{Type: "expire"}},
+				},
+			},
+			want: true,
+		},
+		"ParamSetToUnexpire": {
+			args: args{
+				params: v1alpha1.BucketParameters{
+					Description:    pointer.String("name"),
+					RetentionRules: []v1alpha1.RetentionRule{{Type: "expire"}},
+				},
+				obs: &domain.Bucket{
+					Description: pointer.String("name"),
+				},
+			},
+			want: true,
+		},
+		"ParamNonDefault": {
+			args: args{
+				params: v1alpha1.BucketParameters{
+					Description:    pointer.String("name"),
+					RetentionRules: []v1alpha1.RetentionRule{{Type: "expire", EverySeconds: 3600}},
+				},
+				obs: &domain.Bucket{
+					Description: pointer.String("name"),
+				},
+			},
+			want: false,
+		},
+		"ShardGroupMatch": {
+			args: args{
+				params: v1alpha1.BucketParameters{
+					Description:    pointer.String("name"),
+					RetentionRules: []v1alpha1.RetentionRule{{Type: "expire", ShardGroupDurationSeconds: pointer.Int64(3600)}},
+				},
+				obs: &domain.Bucket{
+					Description:    pointer.String("name"),
+					RetentionRules: []domain.RetentionRule{{Type: "expire", ShardGroupDurationSeconds: pointer.Int64(3600)}},
+				},
+			},
+			want: true,
+		},
+		"ShardGroupUnequal": {
+			args: args{
+				params: v1alpha1.BucketParameters{
+					Description:    pointer.String("name"),
+					RetentionRules: []v1alpha1.RetentionRule{{Type: "expire", ShardGroupDurationSeconds: pointer.Int64(3600)}},
+				},
+				obs: &domain.Bucket{
+					Description:    pointer.String("name"),
+					RetentionRules: []domain.RetentionRule{{Type: "expire", ShardGroupDurationSeconds: pointer.Int64(7200)}},
+				},
+			},
+			want: false,
+		},
+		"RuleLengthCheck": {
+			args: args{
+				params: v1alpha1.BucketParameters{
+					Description:    pointer.String("name"),
+					RetentionRules: []v1alpha1.RetentionRule{{Type: "expire"}},
+				},
+				obs: &domain.Bucket{
+					Description: pointer.String("name"),
+					RetentionRules: []domain.RetentionRule{
+						{Type: "expire"},
+						{Type: "expire", EverySeconds: 3600},
+					},
+				},
+			},
+			want: false,
+		},
+	}
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			res := IsUpToDate(tc.args.params, tc.args.obs)
+
+			if diff := cmp.Diff(tc.want, res); diff != "" {
+				t.Errorf("Observe(...): -want, +got:\n%s", diff)
+			}
+		})
+	}
+}

--- a/package/crds/influxdb.crossplane.io_buckets.yaml
+++ b/package/crds/influxdb.crossplane.io_buckets.yaml
@@ -125,8 +125,6 @@ spec:
                     - implicit
                     - explicit
                     type: string
-                required:
-                - retentionRules
                 type: object
               providerConfigRef:
                 default:

